nerfstudio/fields/tensorf_field.py

```diff
@@ -0,0 +1,138 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""TensoRF Field"""
+
+
+from typing import Optional
+
+import torch
+from torch import nn
+from torch.nn.parameter import Parameter
+from torchtyping import TensorType
+
+from nerfstudio.cameras.rays import RaySamples
+from nerfstudio.data.scene_box import SceneBox
+from nerfstudio.field_components.encodings import Encoding, Identity, SHEncoding
+from nerfstudio.field_components.field_heads import FieldHeadNames, RGBFieldHead
+from nerfstudio.field_components.mlp import MLP
+from nerfstudio.field_components.spatial_distortions import SpatialDistortion
+from nerfstudio.fields.base_field import Field
+
+
+class TensoRFField(Field):
+    """TensoRF Field
+
+    Args:
+        density_encoding:
+        color_encoding:
+
+        head_mlp_num_layers: Number of layer for ourput head MLP.
+        head_mlp_layer_width: Width of output head MLP layers.
+    """
+
+    def __init__(
+        self,
+        aabb,
+        feature_encoding: Encoding = Identity(in_dim=3),
+        direction_encoding: Encoding = Identity(in_dim=3),
+        density_encoding: Encoding = Identity(in_dim=3),
+        color_encoding: Encoding = Identity(in_dim=3),
+        appearance_dim: int = 27,
+        head_mlp_num_layers: int = 2,
+        head_mlp_layer_width: int = 128,
+        spatial_distortion: Optional[SpatialDistortion] = None,
+        use_sh: bool = False,
+        sh_levels: int = 2,
+    ) -> None:
+        super().__init__()
+        self.aabb = Parameter(aabb, requires_grad=False)
+        self.spatial_distortion = spatial_distortion
+        self.feature_encoding = feature_encoding
+        self.direction_encoding = direction_encoding
+        self.density_encoding = density_encoding
+        self.color_encoding = color_encoding
+
+        self.mlp_head = MLP(
+            in_dim=appearance_dim + 3 + self.direction_encoding.get_out_dim() + self.feature_encoding.get_out_dim(),
+            num_layers=head_mlp_num_layers,
+            layer_width=head_mlp_layer_width,
+            activation=nn.ReLU(),
+        )
+
+        self.use_sh = use_sh
+
+        if self.use_sh:
+            self.sh = SHEncoding(sh_levels)
+            self.B = nn.Linear(
+                in_features=self.color_encoding.get_out_dim(), out_features=3 * self.sh.get_out_dim(), bias=False
+            )
+        else:
+            self.B = nn.Linear(in_features=self.color_encoding.get_out_dim(), out_features=appearance_dim, bias=False)
+
+        self.field_output_rgb = RGBFieldHead(in_dim=self.mlp_head.get_out_dim(), activation=nn.Sigmoid())
+
+    def get_density(self, ray_samples: RaySamples):
+        positions = SceneBox.get_normalized_positions(ray_samples.frustums.get_positions(), self.aabb)
+        if self.spatial_distortion is not None:
+            positions = self.spatial_distortion(positions)
```

Comment:
Have you tested with spatial distortions? If not maybe remove it for now. I think there is probably a bug here since the positions are normalized in the line above.

Reply:
Yeah tested with scene contraction and it seems to not affect PSNR but slightly improves rays / sec on the lego scene. Probably not acting as intended with the normalization

Reply:
Yeah tried it with scene contraction and no change to PSNR, probably not doing what it's intended to with the normalization